File: java/perfect-numbers/src/main/java/NaturalNumber.java
final class NaturalNumber {

    private final int number;

    public NaturalNumber(int number) {
        this.number = number;
    }

    public Classification getClassification() {
        int sumOfFactors = 0;

        for (int i = 1; i < number; i++)
            if (number % i == 0)
                sumOfFactors += i;

        if (sumOfFactors < number)
            return Classification.DEFICIENT;
        else if (sumOfFactors == number)
            return Classification.PERFECT;
        else
            return Classification.ABUNDANT;
    }
}

File: kotlin/perfect-numbers/src/main/kotlin/NaturalNumber.kt
enum class Classification {
    DEFICIENT, PERFECT, ABUNDANT
}

fun classify(naturalNumber: Int): Classification {
    require(naturalNumber > 0)

    return naturalNumber.sumOfFactors().let {
        when {
            it < naturalNumber -> Classification.DEFICIENT
            it == naturalNumber -> Classification.PERFECT
            else -> Classification.ABUNDANT
        }
    }
}

fun Int.sumOfFactors() = 1.until(this).filter { this % it == 0 }.sum()